### PR TITLE
8288532: additional review changes for JDK-8286830

### PR DESCRIPTION
**Pull request by dcubed-ojdk**
Update a couple of tests that were modified by JDK-8286830. Now the tests properly
honor any specified time limit by splitting the time between both sub-tests.
Also pick up a sanity check based on JDK-8288497 that I've been using in recent
stress testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288532](https://bugs.openjdk.org/browse/JDK-8288532): additional review changes for JDK-8286830

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [f2baf0b1](https://git.openjdk.org/jdk19/pull/32/files/f2baf0b17b8e278d688594953dd89eaaabb0e4f7)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [f2baf0b1](https://git.openjdk.org/jdk19/pull/32/files/f2baf0b17b8e278d688594953dd89eaaabb0e4f7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk19 pull/32/head:pull/32` \
`$ git checkout pull/32`

Update a local copy of the PR: \
`$ git checkout pull/32` \
`$ git pull https://git.openjdk.org/jdk19 pull/32/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 32`

View PR using the GUI difftool: \
`$ git pr show -t 32`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk19/pull/32.diff">https://git.openjdk.org/jdk19/pull/32.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/thread.inline.hpp [hunk 1]
```diff
@@ -147,6 +147,11 @@ class AsyncExceptionHandshake : public AsyncHandshakeClosure {
   }
 
   ~AsyncExceptionHandshake() {
+    Thread* current = Thread::current();
+    if (current->is_Java_thread()) {
```
Comment (pchilano) [inline src/hotspot/share/runtime/thread.inline.hpp:151]: Why verify if this is a JavaThread? The target is always a JavaThread, and the sender of the exception is also a JavaThread(JVM_StopThread and JvmtiEnv::StopThread). Even the handshake operation will assert if the handshaker is not a JT.

File: test/hotspot/jtreg/runtime/Thread/StopAtExit.java [hunk 2]
```diff
@@ -72,6 +72,8 @@ public static void main(String[] args) {
                 usage();
             }
         }
+        timeMax /= 2;  // Split time between the two sub-tests.
```
Comment (pchilano) [inline test/hotspot/jtreg/runtime/Thread/StopAtExit.java:75]: Maybe bump DEF_TIME_MAX to 60 seconds now so both subtests run the original 30 seconds? My intention was to run the same test again(same time) but with the addition of threadCreator.

File: test/hotspot/jtreg/runtime/Thread/SuspendAtExit.java [hunk 3]
```diff
@@ -78,6 +78,8 @@ public static void main(String[] args) {
                     usage();
             }
         }
+        timeMax /= 2;  // Split time between the two sub-tests.
```
Comment (pchilano) [inline test/hotspot/jtreg/runtime/Thread/SuspendAtExit.java:81]: Same thing about DEF_TIME_MAX.

**Review by pchilano — COMMENTED** [review]
Hi Dan,

Some comments below.

Thanks,
Patricio

File: test/hotspot/jtreg/runtime/Thread/SuspendAtExit.java (hunk 3 above)
Comment (dcubed-ojdk) [reply test/hotspot/jtreg/runtime/Thread/SuspendAtExit.java:81]: I try not to have the default config for a stress test run longer that 30 seconds total.
15 seconds per sub-test is fine.

**Comment by dcubed-ojdk** [comment]
@pchilano - Thanks for the review!

File: src/hotspot/share/runtime/thread.inline.hpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.inline.hpp:151]: Will investigate.

File: test/hotspot/jtreg/runtime/Thread/StopAtExit.java (hunk 2 above)
Comment (dcubed-ojdk) [reply test/hotspot/jtreg/runtime/Thread/StopAtExit.java:75]: I try not to have the default config for a stress test run longer that 30 seconds total.
15 seconds per sub-test is fine.

File: src/hotspot/share/runtime/thread.inline.hpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.inline.hpp:151]: JavaThread::send_async_exception() creates a new AsyncExceptionHandshake.
JavaThread::send_async_exception() is called by JVM_StopThread() and
JvmtiEnv::StopThread(). The InstallAsyncExceptionHandshake which wraps
the AsyncExceptionHandshake is installed via Handshake::execute() which
calls "  JavaThread* self = JavaThread::current()" first thing so send_async_exception()
will fire an assert() if not called by a JavaThread.

Will fix.

**Review by pchilano — APPROVED** [review]
Thanks for the fix. Looks good.

**Comment by dcubed-ojdk** [comment]
I forgot that install_async_exception() can delete the AsyncExceptionHandshake which calls the AsyncExceptionHandshake dtr:
```---------------  T H R E A D  ---------------

Current thread (0x0000ffff001bab70):  VMThread "VM Thread" [stack: 0x0000fffee8110000,0x0000fffee8310000] [id=661857] _threads_hazard_ptr=0x0000fffc8021f770, _nested_threads_hazard_ptr_cnt=0

Stack: [0x0000fffee8110000,0x0000fffee8310000],  sp=0x0000fffee830e030,  free space=2040k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x1909960]  AsyncExceptionHandshake::~AsyncExceptionHandshake()+0x230
V  [libjvm.so+0x1908a7c]  JavaThread::install_async_exception(AsyncExceptionHandshake*)+0x4c
V  [libjvm.so+0x190ace8]  InstallAsyncExceptionHandshake::do_thread(Thread*)+0x38
V  [libjvm.so+0xe67d20]  HandshakeOperation::do_handshake(JavaThread*)+0x13c
V  [libjvm.so+0xe68cec]  HandshakeState::try_process(HandshakeOperation*)+0x23c
V  [libjvm.so+0xe6b384]  VM_HandshakeAllThreads::doit()+0x1f4
V  [libjvm.so+0x19f2788]  VM_Operation::evaluate()+0x168
V  [libjvm.so+0x1a1c9d4]  VMThread::evaluate_operation(VM_Operation*)+0x154
V  [libjvm.so+0x1a1e1bc]  VMThread::inner_execute(VM_Operation*)+0x45c
V  [libjvm.so+0x1a1e374]  VMThread::loop()+0xc4
V  [libjvm.so+0x1a1e4a0]  VMThread::run()+0xcc
V  [libjvm.so+0x1907bb8]  Thread::call_run()+0xf8
V  [libjvm.so+0x16033a4]  thread_native_entry(Thread*)+0x104
C  [libpthread.so.0+0x78f8]  start_thread+0x188
```

**Comment by dcubed-ojdk** [comment]
Gotta back out last change... and restart all the testing... sigh...

**Comment by dcubed-ojdk** [comment]
The latest version's Mach5 Tier[1-7] looks good. Mach5 Tier8 was just started.

@pchilano - please re-review when you get the chance.

File: test/hotspot/jtreg/runtime/Thread/StopAtExit.java [hunk 4]
```diff
@@ -72,6 +72,8 @@ public static void main(String[] args) {
...
+
         test(timeMax);
 
         // Fire-up deamon that just creates new threads. This generates contention on
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Thread/StopAtExit.java:79]: Existing typo: s/deamon/daemon/

**Review by dholmes-ora — APPROVED** [review]
Changes seem fine. It will be nice when the oop-safety check can be handled in OopHandler::release.

Thanks,
David

**Comment by pchilano** [comment]
> I forgot that install_async_exception() can delete the AsyncExceptionHandshake which calls the AsyncExceptionHandshake dtr:
> 
> ```
> 
> Current thread (0x0000ffff001bab70):  VMThread "VM Thread" [stack: 0x0000fffee8110000,0x0000fffee8310000] [id=661857] _threads_hazard_ptr=0x0000fffc8021f770, _nested_threads_hazard_ptr_cnt=0
> 
> Stack: [0x0000fffee8110000,0x0000fffee8310000],  sp=0x0000fffee830e030,  free space=2040k
> Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x1909960]  AsyncExceptionHandshake::~AsyncExceptionHandshake()+0x230
> V  [libjvm.so+0x1908a7c]  JavaThread::install_async_exception(AsyncExceptionHandshake*)+0x4c
> V  [libjvm.so+0x190ace8]  InstallAsyncExceptionHandshake::do_thread(Thread*)+0x38
> V  [libjvm.so+0xe67d20]  HandshakeOperation::do_handshake(JavaThread*)+0x13c
> V  [libjvm.so+0xe68cec]  HandshakeState::try_process(HandshakeOperation*)+0x23c
> V  [libjvm.so+0xe6b384]  VM_HandshakeAllThreads::doit()+0x1f4
> V  [libjvm.so+0x19f2788]  VM_Operation::evaluate()+0x168
> V  [libjvm.so+0x1a1c9d4]  VMThread::evaluate_operation(VM_Operation*)+0x154
> V  [libjvm.so+0x1a1e1bc]  VMThread::inner_execute(VM_Operation*)+0x45c
> V  [libjvm.so+0x1a1e374]  VMThread::loop()+0xc4
> V  [libjvm.so+0x1a1e4a0]  VMThread::run()+0xcc
> V  [libjvm.so+0x1907bb8]  Thread::call_run()+0xf8
> V  [libjvm.so+0x16033a4]  thread_native_entry(Thread*)+0x104
> C  [libpthread.so.0+0x78f8]  start_thread+0x188
> ```
Ah right, the VMThread could take over that synchronous handshake! Thanks for the investigation Dan. The last update looks good.

**Review by pchilano — APPROVED** [review]

**Comment by dcubed-ojdk** [comment]
@pchilano - Thanks for the re-review.
@dholmes-ora - Thanks for the review.

**Comment by dcubed-ojdk** [comment]
The Mach5 Tier8 run had no failures. All testing looks great.

**Pull request closed without merging**